depth: Add methods to derive liquidity allowances on orderbooks by volume and slippage

shazbert opened the pull request:
gctcli: 
* Merge getOrderbookCommand, whaleBombCommand, getOrderbookStreamCommand, getExchangeOrderbookStreamCommand to orderbookcommand
* Link RPCServer to GetOrderbookMovement, GetOrderbookAmountByNominal, GetOrderbookAmountByImpact and add to orderbookcommand

calculator.go:
* Add test and update WhaleBomb() as this was potentially wiping out request price on tranche in either direction. Assign actual impact pct.
* Refactor SimulateOrder() 
* Remove duplication orderSummary ([]Item) and associated methods
* Update buy and sell to DeploymentAction to already include required datatypes for callers

depth.go:

* Add HitTheBids() And LiftTheAsks() methods 
- ByAmount
- ByNominalSlippage
- ByImpactSlippage
- include from mid price reference method
- include best price reference method

* Add supplementary methods GetMidPrice() GetBestBid(), GetBestAsk(), GetSpread() and GetImbalance() 


~~NOTES: Nominal slippage results are not accurate by my calculations, I need help here.~~ It's accurate now.

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -340,6 +340,62 @@ func (ll *linkedList) insertUpdates(updts Items, stack *stack, comp comparison)
+			continue
+		}
+
+		ratio := (initialPrice * (*tip).Value.Price) - 1
```
rn percentageChange

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -340,6 +340,62 @@ func (ll *linkedList) insertUpdates(updts Items, stack *stack, comp comparison)
+		// Full book wiped out, return 100 percent.
+		return 100, nil
+	}
+	ratio := (slippage * shiftedPrice) - 1
```
rn percentageChange or something

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -340,6 +340,62 @@ func (ll *linkedList) insertUpdates(updts Items, stack *stack, comp comparison)
+// getSlippageByVolume returns the slippage percentage by impact volume on
+// liquidity.
+func (ll *linkedList) getSlippageByVolume(volume float64) (float64, error) {
+	var slippage, shiftedPrice float64
```
could return on volume <= 0 or something along those lines to save effort

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -340,6 +340,62 @@ func (ll *linkedList) insertUpdates(updts Items, stack *stack, comp comparison)
+
+// getVolumeBySlippage returns the max volume amount by allowable slippage.
+func (ll *linkedList) getVolumeBySlippage(slippage float64) (float64, error) {
+	var volume, initialPrice float64
```
Could return on slippage <= 0

shazbert replied:
zero percent slippage is the top head tranche I can filter out negative values.

gloriousCode reviewed (changes requested):
Very nice! Good addition

gloriousCode reviewed (commented):
Thanks for making those changes! I just ran a test to make sure I was understanding the calculations correctly. I called `getSlippageByVolume` with a drop in price from 100->90 and I would have expected a slippage value of 10%, however, it returned `9.999999999999998`. The function `(math.CalculatePercentageGainOrLoss(90.0, 100.0)` returns 10.  I know you have done tests on your method of doing the calculations and it is more accurate in some cases, just thought I'd point this out

shazbert commented:
> Thanks for making those changes! I just ran a test to make sure I was understanding the calculations correctly. I called `getSlippageByVolume` with a drop in price from 100->90 and I would have expected a slippage value of 10%, however, it returned `9.999999999999998`. The function `(math.CalculatePercentageGainOrLoss(90.0, 100.0)` returns 10. I know you have done tests on your method of doing the calculations and it is more accurate in some cases, just thought I'd point this out

Yeah, that was a super wrong implementation. I have fixed that now and added more robust tests.

gloriousCode reviewed (approved):
tACK! Thanks for fixing the calculations up

thrasher- commented on exchanges/orderbook/linked_list.go:
```diff
@@ -340,6 +345,74 @@ func (ll *linkedList) insertUpdates(updts Items, stack *stack, comp comparison)
+		// Full book wiped out, return 100 percent.
+		return 100, nil
+	}
+	percentageChange := math.CalculatePercentageGainOrLoss(shiftedPrice, slippage)
```
You should instead be using the average price across different items as per:

```
func TestDepthGetSlippageByVolume2(t *testing.T) {
	t.Parallel()
	d := NewDepth(id)

	asks := Items{
		{
			Price:  10000,
			Amount: 2,
		}, {
			Price:  10100,
			Amount: 7,
		}, {
			Price:  10200,
			Amount: 3,
		},
	}

	d.LoadSnapshot(bid, asks, 0, time.Time{}, true)

	s, err := d.GetSlippageByVolume(10, false)
	if err != nil {
		t.Fatal(err)
	}
	t.Log(s)

	b := Base{
		Asks: asks,
	}
	r1, err := b.GetAveragePrice(true, 10)
	if err != nil {
		t.Fatal(err)
	}

	s2 := (r1 - asks[0].Price) / asks[0].Price
	t.Log(s2 * 100)
}
```

This gives a more accurate slippage value of 0.89% instead of 2%.

I'd also like to be able to toggle between quote and base slippage calculations instead of just base amounts

gloriousCode commented on exchanges/orderbook/calculator.go:
```diff
@@ -15,214 +18,186 @@ type WhaleBombResult struct {
+		action.BaseAmount += b.Bids[x].Amount
 	}
-	return amt, orders
+	return nil, errNotEnoughLiquidity
```
It would be nice to get more information from these responses when liquidity is not enough - to remove guessing games. eg I want to slip to 1337 and I am told "can't be done", so I need to keep modifying values until I know what can be done. I can't even simulate being a whale 😭 

So in that case, is not enough liquidity an error? Or just a different kind of response (which contains an `errNotEnoughLiquidity`)?

gloriousCode commented on exchanges/orderbook/calculator_test.go:
```diff
@@ -27,69 +28,180 @@ func TestWhaleBomb(t *testing.T) {
+		t.Fatalf("received: '%v' but expected: '%v'", result.PercentageGainOrLoss, 0.014285714285714287)
+	}
+
+	fmt.Println(result.Status)
```
no `fmt` in tests please. Use t.Log if you must, but this looks like a leftover

gloriousCode commented on exchanges/orderbook/depth_test.go:
```diff
@@ -624,3 +625,836 @@ func TestIsValid(t *testing.T) {
+	}
+
+	// First and second price
+	amt, err := depth.GetBaseFromNominalSlippage(0.018443378827001, 1336)
```
My own personal note. I need to understand what's happening when I'm inputting a price of 3000

shazbert replied:
FYI: The reference price in this case can be set e.g. TWAP, MID or BEST price etc. It was for a future consideration when we don't want to fall by a great degree from a certain point while deploying capital to a book.

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -394,6 +526,108 @@ func (ll *asks) insertUpdates(updts Items, stack *stack) error {
 	return ll.linkedList.insertUpdates(updts, stack, askCompare)
 }
 
+// getMovementByQuoteAmount lifts ask liquidy with quotation amount to acquire
```
```suggestion
// getMovementByQuoteAmount lifts ask liquidity with quotation amount to acquire
```

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -526,3 +760,42 @@ func shiftBookmark(tip *Node, bookmark, head **Node, updt Item) bool {
+		return nil, errAmountExceedsSideLiquidity
+	}
+	averageOrderPrice := totalValue / amounts
+	// Nominal pct deference is from the reference price twap/vwap, mid, best
```
difference?

gloriousCode reviewed (changes requested):
This PR does not resemble the initial one! 📈 This is just an initial review, I'll need to be doing more indepth testing and understanding of all the additional functionality you've added

gloriousCode commented on cmd/gctcli/orderbook.go:
```diff
@@ -0,0 +1,744 @@
+}
+
+var quote = &cli.Command{
+	Name:      "quote",
```
Another comment on this one. Is it possible to infer pricing? I'll run you through what I'm getting at.
I performed the following command:
```
go run . orderbook whalebomb --exchange="ftx" --asset="futures" --pair="btc-perp" --price="21790" --side="buy"

 "orders": [
   ...
 ],
 "amount": 3737841.4552,
 "minimum_price": 21774,
 "maximum_price": 21790,
 "percentage_gain_loss": 0.07348213465601176,
 "status": "Buying 3737841.46 PERP worth of BTC will send the price from 21774 to 21790 [0.07%] and take 16 orders. "
}
```
From here, it tells me, in order to move the price, I need to spend `3737841.4552`
So then I do the following:
```
go run . orderbook buy quote ftx btc-perp futures 3737841.4552
{
 "nominal_percentage": 0.03983502546274574,
 "impact_percentage": 0.07348213465601176,
 "slippage_cost": 1488.3771999995224,
 "currency_bought": "BTC",
 "currency_sold": "PERP",
 "side_affected": "ASK",
 "update_protocol": "REST"
}
```

What I see here, is that the `impact_percentage` matches the `percent_gain_loss`. Hurray! :tada: , but what is missing, is the price that it roughly lands at, which is about 21790 `PERP`. Is the pricing something that is known and can be added? Or, am I misusing this command, and as a result, should just stick to whalebomb? Because as the description says, this command is simulating a buy using quote, so it could theoretically, say where the price has gone as a result of this buy (which I believe is the impact percentage)

gloriousCode commented on exchanges/orderbook/calculator_test.go:
```diff
@@ -27,69 +29,178 @@ func TestWhaleBomb(t *testing.T) {
+	}
+}
+
+func TestSimulateOrder(t *testing.T) {
```
Are you able to extend coverage of this to have successful results?

gloriousCode commented on cmd/gctcli/orderbook.go:
```diff
@@ -0,0 +1,744 @@
+		price = c.Float64("price")
+	} else if c.Args().Get(4) != "" {
+		var err error
+		price, err = strconv.ParseFloat(c.Args().Get(3), 64)
```
```suggestion
		price, err = strconv.ParseFloat(c.Args().Get(4), 64)
```

gloriousCode reviewed (changes requested):
This is looking nice! I'm going to need to verify the percentage calculations in a future review :nerd_face:

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -526,3 +822,42 @@ func shiftBookmark(tip *Node, bookmark, head **Node, updt Item) bool {
 }
+
+// getMovement returns movement on tranche amounts
+func getMovement(leftover, amounts, totalValue, headPrice, refPrice, tranchePrice float64, noLiquidity bool) (*Movement, error) {
```
I want to work with you on this soon. But I'm looking at this function to emulate [this](https://en.wikipedia.org/wiki/Slippage_(finance)) example.
Essentially calling:
```go
	movement, err := getMovement(0, 1538.46153846154, 232485.923076923, 151.08, 151.08, 151.25, false)

```
Where 1,538.46153846154 is avg order size, 232,485.923076923 is avg purchase cost, 151.08 is head price and trancheprice (aka final price) is 151.25.
I can see that your averageOrderPrice is correct :tada: but I'm having trouble getting the -717 slippage cost. 

If I wanted to get the slippageCost as per the example, I would need to have the following:
```go
averageOrderPrice := totalValue / amounts
quantityPurchased := 20000
slippageCost := (refPrice * quantityPurchased) - (averageOrderPrice * quantityPurchased)
```
So based on that, this function is missing the total quantity purchased in order to work out this slippage cost. But it seems like you have a different method of calculating is, so that's why I want to work with you to understand what you've done versus the referenced example

gloriousCode commented on exchanges/orderbook/linked_list.go:
```diff
@@ -526,3 +822,42 @@ func shiftBookmark(tip *Node, bookmark, head **Node, updt Item) bool {
+	}
+
+	m := &Movement{
+		NominalPercentage:    nominalP,
```
I think `averageOrderPrice` would be a lovely property to include in `Movement`

gloriousCode reviewed (changes requested):
Thanks for making those changes! Most of these comments have been from testing your work against [wikipedia](https://en.wikipedia.org/wiki/Slippage_(finance))

gloriousCode reviewed (changes requested):
I can now see that the slippage calculation is correct :tada: Nice!
```
NominalPercentage : 0.023729
ImpactPercentage : -100.000000
SlippageCost : 717.000000
CurrencyBought : BTC
Bought : 20000.000000
CurrencySold : USD
Sold : 3022317.000000
SideAffected : ASK
UpdateProtocol : REST
FullOrderbookSideConsumed : true
StartPrice : 151.080000
EndPrice : 0.000000
NoSlippageOccurred : false
AverageOrderCost : 151.115850
```
After these, I think it will be one last check and good to go

gloriousCode commented on exchanges/orderbook/depth.go:
```diff
@@ -323,10 +279,361 @@ func (d *Depth) IsFundingRate() bool {
+func (d *Depth) TotalAskAmounts() (liquidity, value float64, err error) {
+	d.m.Lock()
+	defer d.m.Unlock()
+	if d.validationError != nil {
```
Is it okay for the new functions below to be lacking this validation error check?

shazbert replied:
No not at all! 😭

gloriousCode reviewed (changes requested):
Just two basic ones. This is looking pretty swell

shazbert commented on engine/rpcserver.go:
```diff
@@ -5327,6 +5327,15 @@ func (s *RPCServer) GetOrderbookMovement(ctx context.Context, r *gctrpc.GetOrder
 		return nil, err
 	}
 
+	if pair.IsEmpty() {
+		return nil, currency.ErrCurrencyPairEmpty
+	}
```
Yeah it returns nil, I wanted it to complain at the RPC level and not when I fetch from the orderbook system.

gloriousCode replied:
I did not notice that `nil` return! :tada:

gloriousCode reviewed (approved):
tACK! Nice work Shazbert 
# :sun_with_face: :rainbow:

thrasher- commented on exchanges/orderbook/linked_list.go:
```diff
@@ -3,11 +3,28 @@ package orderbook
+	errCollisionDetected               = errors.New("cannot insert update collision detected")
+	errAmountCannotBeLessOrEqualToZero = errors.New("amount cannot be less or equal to zero")
+	errInvalidNominalSlippage          = errors.New("invalid slippage amount must be greater or equal to zero")
+	errInvalidImpactSlippage           = errors.New("invalid slippage amount must be greater than zero")
```
```suggestion
	errInvalidImpactSlippage           = errors.New("invalid slippage amount, its value must be greater than zero")
```

thrasher- commented on exchanges/orderbook/linked_list.go:
```diff
@@ -3,11 +3,28 @@ package orderbook
+	errAmountCannotBeLessOrEqualToZero = errors.New("amount cannot be less or equal to zero")
+	errInvalidNominalSlippage          = errors.New("invalid slippage amount must be greater or equal to zero")
+	errInvalidImpactSlippage           = errors.New("invalid slippage amount must be greater than zero")
+	errInvalidSlippageCannotExceed100  = errors.New("invalid slippage cannot exceed 100%")
```
```suggestion
	errInvalidSlippageCannotExceed100  = errors.New("invalid slippage amount, its value cannot exceed 100%")
```

thrasher- reviewed (changes requested):
Minor grammatical/consistency nits. tACK all commands including trying to break them and thanks for enhancing it heavily based on my first nit. A very nice addition to OB functionality!

thrasher- reviewed (approved):
thanks for making those changes!
